miden-lib/asm/miden/kernels/tx/tx.masm

```diff
@@ -107,10 +116,144 @@ proc.increment_num_created_notes
+    # => [asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # compute the pointer at which we should stop iterating
+    dup dup.7 add
+    # => [end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # pad the stack and move the pointer to the top
+    padw movup.5
+    # => [asset_ptr, 0, 0, 0, 0, end_asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # compute the loop latch
+    dup dup.6 neq
+    # => [latch, asset_ptr, 0, 0, 0, 0, end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    while.true
+        dup movdn.6
+        # => [asset_ptr, 0, 0, 0, 0, end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+        # load the asset and compare
+        mem_loadw push.0 movdn.4 push.0 movdn.4 dupw.2
+        # => [ASSET, ASSET', 0, 0, end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+
+        # compare the ASSETs
+        eqw assertz.err=ERR_NON_FUNGIBLE_ASSET_ALREADY_EXISTS
+        # => [ASSET, ASSET', 0, 0, end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+
+        # drop ASSETs and increment the asset pointer
+        dropw dropw push.0.0 movup.5 add.1
+        # => [asset_ptr, 0, 0, 0, 0, end_asset_ptr, ASSET, note_ptr, num_of_assets]
+
+        # check if we reached the end of the loop
+        dup dup.6 neq
+    end
+    # => [asset_ptr, 0, 0, 0, 0, end_asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # prepare stack for storing the ASSET
+    drop dropw
+    # => [end_asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # end of the loop reached, no error so we can store the non-fungible asset
+    mem_storew dropw
+    # => [note_ptr, num_of_assets]
+
+    # increase the number of assets in the note
+    swap add.1 dup.1 exec.memory::set_created_note_num_assets
+    # => [note_ptr]
+end
+
+#! Adds a fungible asset to a note. If the note already holds an asset issued by the
+#! same faucet id the two quantities are summed up and the new quantity is stored at the
+#! old position in the note. In the other case, the asset is stored at the next available
+#! position. Returns the pointer to the note the asset was stored at.
+#! Panics if the combined quantity exceeds the maximum for fungible assets (~overflow).
+#!
+#! Inputs: [ASSET, note_ptr, num_of_assets]
+#! Outputs: [note_ptr]
+proc.add_fungible_asset_to_note
+    dup.4 exec.memory::get_created_note_asset_data_ptr
+    # => [asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # compute the pointer at which we should stop iterating
+    dup dup.7 add
+    # => [end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # pad the stack and move the pointer to the top
+    padw movup.5
+    # => [asset_ptr, 0, 0, 0, 0, end_asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    # compute the loop latch
+    dup dup.6 neq
+    # => [latch, asset_ptr, 0, 0, 0, 0, end_asset_ptr, ASSET, note_ptr, num_of_assets]
+
+    while.true
+        dup movdn.6
+        # => [asset_ptr, 0, 0, 0, 0, end_asset_ptr, asset_ptr, ASSET, note_ptr, num_of_assets]
```

Comment:
Similar to the previous comment: I think if we enter the loop with `ASSET` being closer to the top of the stack, we can make things more efficient. Specifically, it could look like so:
```
[asset_ptr, 0, 0, 0, 0, ASSET, end_asset_ptr, asset_ptr, note_ptr, num_of_assets]
```
And then we immediately do:
```
mem_loadw
# => [STORED_ASSET, ASSET, end_asset_ptr, asset_ptr, note_ptr, num_of_assets]

dup.4 eq
# => [are_equal, 0, 0, stored_amount, ASSET, end_asset_ptr, asset_ptr, note_ptr, num_of_assets]

if.true
    movup.2 movup.6 add
    # => [updated_amount, 0, 0, faucet_id, 0, 0, end_asset_ptr, asset_ptr, note_ptr, num_of_assets]
    ...

    # => [asset_ptr, 0, 0, 0, 0, ASSET', end_asset_ptr, asset_ptr, note_ptr, num_of_assets]
    push.0
else
    push.0 movup.9 add.1 dup movdn.10
    # => [asset_ptr+1, 0, 0, 0, stored_amount, ASSET, end_asset_ptr, asset_ptr+1, note_ptr, num_of_assets]

    dup.9 dup.1 neq
    # => [latch, asset_ptr+1, 0, 0, 0, stored_amount, ASSET, end_asset_ptr, asset_ptr+1, note_ptr, num_of_assets]
end
```
And here also we'd avoid a bunch of drops and dups.